src/k8s/pkg/component/dns.go

```diff
@@ -132,3 +134,130 @@ func DisableDNSComponent(ctx context.Context, s snap.Snap) error {
+					{"name": "cache", "parameters": "30"},
+					{"name": "loop"},
+					{"name": "reload"},
+					{"name": "loadbalance"},
+				},
+			},
+		},
+	}
+
+	if serviceIP != "" {
+		service := values["service"].(map[string]any)
+		service["clusterIP"] = serviceIP
+	}
+
+	if isRefresh {
+		if err := manager.Refresh("dns", values); err != nil {
+			return "", "", fmt.Errorf("failed to refresh dns component: %w", err)
+		}
+	} else {
+		if err := manager.Enable("dns", values); err != nil {
+			return "", "", fmt.Errorf("failed to enable dns component: %w", err)
+		}
+	}
+
+	client, err := k8s.NewClient(s)
+	if err != nil {
+		return "", "", fmt.Errorf("failed to create kubernetes client: %w", err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	dnsIP, err := client.GetServiceClusterIP(ctx, "coredns", "kube-system")
+	if err != nil {
+		return "", "", fmt.Errorf("failed to get dns service: %w", err)
+	}
+
+	// TODO: this should propagate to all nodes
+	changed, err := snaputil.UpdateServiceArguments(s, "kubelet", map[string]string{
+		"--cluster-dns":    dnsIP,
+		"--cluster-domain": clusterDomain,
+	}, nil)
+	if err != nil {
+		return "", "", fmt.Errorf("failed to update kubelet arguments: %w", err)
+	}
+
+	if changed {
+		ctx, cancel = context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+
+		if err := s.RestartService(ctx, "kubelet"); err != nil {
+			return "", "", fmt.Errorf("failed to restart kubelet to apply new dns configuration: %w", err)
+		}
+	}
+	return dnsIP, clusterDomain, nil
+}
+
+func ReconcileDNSComponent(ctx context.Context, s snap.Snap, alreadyEnabled *bool, requestEnabled *bool, clusterConfig types.ClusterConfig) (string, string, error) {
+	if (requestEnabled == nil && alreadyEnabled != nil && *alreadyEnabled) || (requestEnabled != nil && *requestEnabled && alreadyEnabled != nil && *alreadyEnabled) {
+		// If already enabled, and request does not contain `enabled` key
+		// or if already enabled and request contains `enabled=true`
+		dnsIP, clusterDomain, err := UpdateDNSComponent(ctx, s, true, clusterConfig.Kubelet.ClusterDomain, clusterConfig.Kubelet.ClusterDNS, clusterConfig.DNS.UpstreamNameservers)
+		if err != nil {
+			return "", "", fmt.Errorf("failed to refresh dns: %w", err)
+		}
+		return dnsIP, clusterDomain, nil
+	} else if requestEnabled != nil && *requestEnabled {
+		dnsIP, clusterDomain, err := UpdateDNSComponent(ctx, s, false, clusterConfig.Kubelet.ClusterDomain, clusterConfig.Kubelet.ClusterDNS, clusterConfig.DNS.UpstreamNameservers)
+		if err != nil {
+			return "", "", fmt.Errorf("failed to enable dns: %w", err)
+		}
+		return dnsIP, clusterDomain, nil
+	} else if requestEnabled != nil && !*requestEnabled {
+		err := DisableDNSComponent(ctx, s)
+		if err != nil {
+			return "", "", fmt.Errorf("failed to disable dns: %w", err)
+		}
+		return "", "", nil
+	} else {
+		// do nothing
+	}
+	return "", "", nil
+}
```

Comment:
I see why we do it like so, but we probably want to revisit this logic in the future

Reply:
I am not exactly happy about this as well, the main reason is getting the assigned Service IP to persist it.